ml-agents/mlagents/trainers/settings.py

```diff
@@ -312,6 +294,168 @@ def apply(self, key: str, env_channel: EnvironmentParametersChannel) -> None:
         )
 
 
+# ENVIRONMENT PARAMETERS ###############################################################
+@attr.s(auto_attribs=True)
+class CompletionCriteriaSettings:
+    """
+    CompletionCriteriaSettings contains the information needed to figure out if the next
+    lesson must start.
+    """
+
+    class MeasureType(Enum):
+        PROGRESS: str = "progress"
+        REWARD: str = "reward"
+
+    measure: MeasureType = attr.ib(default=MeasureType.REWARD)
+    behavior: str = attr.ib(default="")
+    min_lesson_length: int = 0
+    signal_smoothing: bool = True
+    threshold: float = attr.ib(default=0.0)
+    require_reset: bool = False
+
+    @threshold.validator
+    def _check_threshold_value(self, attribute, value):
+        """
+        Verify that the threshold has a value between 0 and 1 when the measure is
+        PROGRESS
+        """
+        if self.measure == self.MeasureType.PROGRESS:
+            if self.threshold > 1.0:
+                raise TrainerConfigError(
+                    "Threshold for next lesson cannot be greater than 1 when the measure is progress."
+                )
+            if self.threshold < 0.0:
+                raise TrainerConfigError(
+                    "Threshold for next lesson cannot be greater negative when the measure is progress."
+                )
+
+
+@attr.s(auto_attribs=True)
+class Lesson:
+    """
+    Gathers the data of one lesson for one environment parameter including its name,
+    the condition that must be fullfiled for the lesson to be completed and a sampler
+    for the environment parameter. If the completion_criteria is None, then this is
+    the last lesson in the curriculum.
+    """
+
+    completion_criteria: Optional[CompletionCriteriaSettings]
+    sampler: ParameterRandomizationSettings
+    name: str
+
+
+@attr.s(auto_attribs=True)
+class EnvironmentParameterSettings:
+    """
+    EnvironmentParameterSettings is an ordered list of lessons for one environment
+    parameter.
+    """
+
+    lessons: List[Lesson]
```

Comment:
I think this generally should match the YAML structure/keys, otherwise it won't be able to be dumped. So in this case this class should contain `curriculum: List[Lesson]`.

Reply:
I create these using the structure method, not the default converter. This code works as intended.
I did not name it curriculum because it can sometimes contain a single lesson (not a curriculum wen there is a single value).

Reply:
IMO it's kind of defeating the purpose of the `attrs` classes if they don't match the YAML. In the other classes doing an `unstructure` of the class structure creates a usable YAML file. 

Hence I think the `environment_parameters` dict should contain as the entries either this object (which contains `curriculum`) or a `Lesson` object if it only has a single lesson.

Reply:
I think I see what you are saying. And I think it is related to [your other comment](https://github.com/Unity-Technologies/ml-agents/pull/4160#discussion_r445261976). I am trying to make the EnvironmentParameterSetting agnostic of how the parameters where written in the yaml. This allows me to have much simpler logic in the `update_lesson` method since I know I always have a list of lessons, always with a completion criteria and a sampler. But in the yaml, there are multiple ways to specify a sampler : 
```yaml
small_wall_height:
    curriculum:
    - Lesson0:
        value: 1.5
small_wall_height: 1.5
other_wall_height:
  sampler_type: uniform
  sampler_parameters:
    min_value: 1.50
    max_value: 1.51
```
But all must result in a list of lessons. I could have the settings mirror the yaml, but then I would need to write a converter from the settings to the list of Lessons.
Do you think there is a way to both mirror the yaml and convert it to a list of lessons ?
Or do you think I should not have a list of lessons to begin with and have my logic be case by case (curriculum / sampler / constant) ?